lms/djangoapps/instructor_task/tasks_helper/misc.py

```diff
@@ -340,3 +349,227 @@ def upload_ora2_data(
+
+    try:
+        return function_to_run()
+    # Update progress to failed regardless of error type
+    except Exception:  # pylint: disable=broad-except
+        TASK_LOG.exception(exception_text)
+        task_progress.failed = 1
+
+        task_progress.update_task_state(extra_meta={'step': step_error_description})
+
+
+def _create_downloads_csv(submission_files_data):
+    """
+    Creates csv where each row represents all information about some certain
+    submission attachment.
+    """
+
+    header = (
+        'Block Id',
+        'Student Id',
+        'File Key',
+        'File Name',
+        'File Description',
+        'File Size',
+        'Path',
+    )
+
+    rows = []
+
+    for block_id, submissions in submission_files_data.items():
+        for student_id, submission_files in submissions.items():
+            for attachment in submission_files['attachments']:
+                path_of_attachment_in_zip = os.path.join(
+                    block_id,
+                    student_id,
+                    'attachments',
+                    attachment['name'],
+                )
+
+                rows.append(
+                    (
+                        block_id,
+                        student_id,
+                        attachment['key'],
+                        attachment['name'],
+                        attachment['description'],
+                        attachment['size'],
+                        path_of_attachment_in_zip,
+                    )
+                )
+
+    output_buffer = StringIO()
+    csvwriter = csv.writer(output_buffer)
+    csvwriter.writerows([header] + rows)
+
+    return output_buffer
+
+
+def _create_zip_with_attachments(file, course_id, submission_files_data):
+    """
+    Opens given stream as a zip file and writes into it all submission
+    attachments and csv with list of all downloads.
+
+    Example of result zip file structure:
+    ```
+    .
+    └── CourseId
+        ├── BlockId1
+        │   ├── StudentId1
+        │   │   ├── attachments
+        │   │   │   ├── SomeFile1
+        │   │   │   └── SomeFile2
+        │   │   ├── part_0.txt
+        │   │   └── part_1.txt
+        │   └── StudentId2
+        │       ├── part_0.txt
+        │       └── part_1.txt
+        ├── BlockId2
+        │   └── StudentId3
+        │       ├── attachments
+        │       │   └── SomeFile4
+        │       └── part_0.txt
+        └── downloads.csv
+    ```
+    """
+
+    storage = DefaultStorage()
+
+    with ZipFile(file, 'w') as zip_file:
+        for block_id, submissions in submission_files_data.items():
+            for student_id, submission_files in submissions.items():
+                for text in submission_files['texts']:
+                    text_path = os.path.join(
+                        str(course_id),
+                        block_id,
+                        student_id,
+                        text['filename'],
+                    )
```

Comment:
nit: You can save this base path per student id and just append text['filename'] to it. It will clarify the code a bit.